### DISCUSSION TITLE
UG Loopback for NAT HolePunching

### DISCUSSION THREAD

**Discussion by sogorman** [CESNET/UltraGrid#353] [General]
Love the Nat hole punching, it's been great. The biggest issue we are having now is wanting to send one encode stream to multipole UG receivers.

Is something like this possible... use one UG instance as a main Encode Process taking in a feed (in this case just a testcard) and preforming the main encode and sending it back to the local host
```
uv.exe -t testcard -c libavcodec:codec=H.265:bitrate=1M --audio-filter controlport_stats -s testcard --audio-capture-format channels=2 --audio-codec MP3:bitrate=192K --control-port 0 -P 10000 127.0.0.1 --param errors-fatal
```

On that same host then use two UG instances to NAT hole punch to two separate recievers

```
uv.exe -s embedded -t ug_input:10000 -c none  -Nholepunch:room=TESTROOM2:server=xxx.xxx.xxx.xxx
```

```
uv.exe -s embedded -t ug_input:10000 -c none  -Nholepunch:room=TESTROOM1:server=xxx.xxx.xxx.xxx
```

I can start the main encode process and one of the secondary nat senders but when starting the second nat sender, UG just sits on 

```
UltraGrid 1.8.2 (tags/v1.8.2 rev f032226 built Mar 21 2023 14:19:43)

Display device   : none
Capture device   : ug_input
Audio capture    : embedded
Audio playback   : none
MTU              : 1500 B
Video compression: none
Audio codec      : PCM
Network protocol : UltraGrid RTP
Audio FEC        : none
Video FEC        : none

Audio receiving started.
Control socket listening on port 25825
Audio sending started.

```

**Comment by alatteri** [comment]
Use the reflector (hd_rum_transcode) for that.  But not sure if it also does hole-punch.

**Reply by sogorman** [reply]
Agree hd_rum_transcode is the correct non hole punch solution but I don't believe hd_rum_transcode supports NAT punching

**Reply by alatteri** [reply]
We run all UG operations inside a ZeroTier VPN, encoders and clients.  This essentially gets you free hole-punch at the base network layer, so you can do creative implementation without restriction.
